### ISSUE TITLE
clangd code completion works randomly!

### ISSUE THREAD

**Issue by atishp04** [FlorentRevest/linux-kernel-vscode#2]
@FlorentRevest 
Thanks for your work. The vscode setup is really amazing. I added a patch for RISC-V.
https://github.com/atishp04/linux-kernel-vscode/commit/6e0b6238827edd8d64d1ea16b51921d21e1e7412

I am planning to send a PR but I face code completion/navigation issues randomly on my setup. I am not sure if it is specific cross compilation or something else. My linux kernel source lives on a remote VM and I use the vscode remote explorer plugin to connect to remote machine to browse/compile the source.

The setup is perfect when it works but the code completion/navigation stops randomly. Disabling/enabling extensions works for some time but stops again after a while. Any thoughts ?

**Comment by FlorentRevest** [comment]
> @FlorentRevest Thanks for your work. The vscode setup is really amazing. I added a patch for RISC-V. [atishp04@6e0b623](https://github.com/atishp04/linux-kernel-vscode/commit/6e0b6238827edd8d64d1ea16b51921d21e1e7412)

Awesome! That looks fairly nice yes! :) 

The PR would be very welcome, just a couple of comments:
- it would be great if you could try the `chroot` task with `  : ${PROOT_ARGS:="-q qemu-riscv64-static"}` as well
- I see you use `riscv64-unknown-linux-gnu` while we currently use `aarch64-linux-gnu` on arm64 for example. It seems to me we should keep those consistent. If there is a good reason for adding -unknown to the target triplet that sounds good to me. Otherwise, maybe just using `riscv64-linux-gnu` :) 

Also, I see you keep a local modification of tasks.sh to disable silent building: https://github.com/atishp04/linux-kernel-vscode/commit/6353fc709b5b7126389f06a4a8f9eaabdd5c0ded FWIW, the philosophy of this script is that these modifications can be toggled from `local.sh` so that `tasks.sh` can always be overwritten. It should be a simple matter of uncommenting this line instead: https://github.com/atishp04/linux-kernel-vscode/blob/main/local.sh#L26

> I am planning to send a PR but I face code completion/navigation issues randomly on my setup. I am not sure if it is specific cross compilation or something else. My linux kernel source lives on a remote VM and I use the vscode remote explorer plugin to connect to remote machine to browse/compile the source.

I use the remote setup all the time as well and I don't think that would interact with the problem here.

> The setup is perfect when it works but the code completion/navigation stops randomly. Disabling/enabling extensions works for some time but stops again after a while. Any thoughts ?

Similar bugs have been reported to me by other users but somehow I don't encounter these problems so I'm unable to debug them :/ Anyway it seems like it would firmly be a bug in clangd or the clangd vscode extension and they would be in a better position to help you with those! :) https://github.com/clangd/vscode-clangd/issues

**Comment by atishp04** [comment]
> > @FlorentRevest Thanks for your work. The vscode setup is really amazing. I added a patch for RISC-V. [atishp04@6e0b623](https://github.com/atishp04/linux-kernel-vscode/commit/6e0b6238827edd8d64d1ea16b51921d21e1e7412)
> 
> Awesome! That looks fairly nice yes! :)
> 
> The PR would be very welcome, just a couple of comments:
> 
> * it would be great if you could try the `chroot` task with `  : ${PROOT_ARGS:="-q qemu-riscv64-static"}` as well
> * I see you use `riscv64-unknown-linux-gnu` while we currently use `aarch64-linux-gnu` on arm64 for example. It seems to me we should keep those consistent. If there is a good reason for adding -unknown to the target triplet that sounds good to me. Otherwise, maybe just using `riscv64-linux-gnu` :)
> 

Yes. I am yet to try the chroot task. I really like the debugging through the VM via VScode. I will update my branch after I test that.

> Also, I see you keep a local modification of tasks.sh to disable silent building: [atishp04@6353fc7](https://github.com/atishp04/linux-kernel-vscode/commit/6353fc709b5b7126389f06a4a8f9eaabdd5c0ded) FWIW, the philosophy of this script is that these modifications can be toggled from `local.sh` so that `tasks.sh` can always be overwritten. It should be a simple matter of uncommenting this line instead: https://github.com/atishp04/linux-kernel-vscode/blob/main/local.sh#L26
> 

Yeah. That won't be part of the MR. That's just my preference. 

> > I am planning to send a PR but I face code completion/navigation issues randomly on my setup. I am not sure if it is specific cross compilation or something else. My linux kernel source lives on a remote VM and I use the vscode remote explorer plugin to connect to remote machine to browse/compile the source.
> 
> I use the remote setup all the time as well and I don't think that would interact with the problem here.
> 

Good to know. There is hope. I thought there is something fundamentally broken with remote explorer.

> > The setup is perfect when it works but the code completion/navigation stops randomly. Disabling/enabling extensions works for some time but stops again after a while. Any thoughts ?
> 
> Similar bugs have been reported to me by other users but somehow I don't encounter these problems so I'm unable to debug them :/ Anyway it seems like it would firmly be a bug in clangd or the clangd vscode extension and they would be in a better position to help you with those! :) https://github.com/clangd/vscode-clangd/issues

ok. I am not the only one with wrong setup. I will continue to hunt for the root cause.

Another question:
Have you tried building kselftests on ARM64 with the clangd ? For some reason, it doesn't work on RISC-V. The header gets resolved to the host. Here is an example

`make O=/tmp/kvm_selftests ARCH=riscv CROSS_COMPILE=riscv64-unknown-linux-gnu- CC='ccache clang' LLVM=1 LLVM_IAS=1 EXTRA_CFLAGS=-static -j 31 -C tools/testing/selftests/kvm                                                                 
make: Entering directory '/scratch/workspace/upstream/linux/tools/testing/selftests/kvm'                                                                                                                                                      
ccache clang  -Wall -Wstrict-prototypes -Wuninitialized -O2 -g -std=gnu99 -Wno-gnu-variable-sized-type-not-at-end -MD -MP -DCONFIG_64BIT -fno-builtin-memcmp -fno-builtin-memcpy -fno-builtin-memset -fno-builtin-strnlen -fno-stack-protector
 -fno-PIE -I/scratch/workspace/upstream/linux/tools/testing/selftests/../../../tools/include -I/scratch/workspace/upstream/linux/tools/testing/selftests/../../../tools/arch/riscv/include -I/scratch/workspace/upstream/linux/tools/testing/s
elftests/../../../usr/include/ -Iinclude -Iriscv -Iinclude/riscv -I ../rseq -I.. -static -isystem /scratch/workspace/upstream/linux/tools/testing/selftests/../../../usr/include   -c riscv/sbi_pmu.c -o /scratch/workspace/upstream/linux/too
ls/testing/selftests/kvm/riscv/sbi_pmu.o                                                                                                                                                                                                      
In file included from riscv/sbi_pmu.c:16:                                                                                                                                                                                                     
In file included from include/kvm_util.h:10:                                                                                                                                                                                                  
In file included from include/kvm_util_base.h:20:                                                                                                                                                                                             
In file included from /scratch/workspace/upstream/linux/tools/testing/selftests/../../../tools/include/asm/atomic.h:6:                                                                                                                        
/scratch/workspace/upstream/linux/tools/testing/selftests/../../../tools/include/asm/../../arch/x86/include/asm/atomic.h:11:10: fatal error: 'asm/asm.h' file not found                                                                       
#include <asm/asm.h>                                                                                                                                                                                                                          
         ^~~~~~~~~~~                                                                                                                                                                                                                          
1 error generated.                                                                                                                                                                                                                            
make: *** [Makefile:283: /scratch/workspace/upstream/linux/tools/testing/selftests/kvm/riscv/sbi_pmu.o] Error 1                                                                                                                               
make: Leaving directory '/scratch/workspace/upstream/linux/tools/testing/selftests/kvm' `
